### PR TITLE
Fix memory-safety and correctness bugs surfaced by Coverity audit (part 7)

### PR DESCRIPTION
**Pull request by stelfrag**
##### Summary
- Split / based on  #22234

---
## Summary by cubic
Fixes ML locking and training edge cases to keep host stats coherent, avoid long mutex holds, serialize start/stop, and make k-means deterministic and safe with undersized inputs.

- **Bug Fixes**
  - Detection/host stats: aggregate locally, snapshot/compare ml_stop_generation around the unlocked chart walk, and publish host->mls and host_anomaly_rate under host->mutex; on ml_running=false or generation change, discard the snapshot, zero per-context anomaly counters, and republish. Addresses CIDs 410121, 393056.
  - Start/stop synchronization: add host->start_stop_mutex to serialize ml_host_start() and ml_host_stop(); stop holds it across the full reset and the final generation bump, start holds it during its sweep; prevents re-enabling mid-reset and torn snapshots.
  - Stop path: reset host stats under host->mutex, release it before chart/dim traversals, and bump ml_stop_generation only after all resets commit; initialize it to 0 in ml_host_new(). Addresses CID 393057.
  - ML training correctness: early-return before k-means when preprocessing yields fewer than 2 vectors; under dim->slock call ml_dimension_finalize_constant_state() (mt=CONSTANT, ts=TRAINED, reset suppression), and clear training_in_progress.
  - K-means model safety: zero-initialize ml_kmeans_inlined_t centers by default and when converting/assigning from an empty model; preserve min/max and window bounds. Addresses CID 451560.
  - Config guard: clamp max_training_vectors to [2, 86400].
  - Tests: add unit tests for empty-model conversion, the 1-vector preprocess boundary, and finalize-constant-state.
  - Minor: simplify context_anomaly_rate iteration (no behavior change).

<sup>Written for commit 65817c6515c780500cd40a7ce65f75eca775f093. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/netdata/netdata/pull/22280?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by stelfrag** [comment]
@cubic-dev-ai rewrite the summary in the PR description

**Comment by stelfrag** [comment]
@cubic-dev-ai write summary

File: src/ml/ml.cc [hunk 1]
```diff
@@ -1032,20 +1051,20 @@ ml_host_detect_once(ml_host_t *host, ONEWAYALLOC *owa)
 
             ml_machine_learning_stats_t chart_mls = chart->mls;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/ml/ml.cc:1118]: Will handle in another PR

File: src/ml/ml_public.cc [hunk 2]
```diff
@@ -177,6 +182,10 @@ void ml_host_stop(RRDHOST *rh) {
     host->mls = ml_machine_learning_stats_t();
     ml_host_clear_context_anomaly_rate(host);
 
+    // Chart deletion can hold the dictionary writer across lengthy cleanup.
+    // Do not carry host->mutex into the traversal below.
+    netdata_mutex_unlock(&host->mutex);
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/ml/ml_public.cc:200]: Fixed in 65817c6515: added a dedicated start_stop_mutex to ml_host_t. ml_host_stop() now holds it across the entire stop sequence (chart/dim reset walk + final generation bump) and ml_host_start() acquires it before re-enabling ml_running, so a start can no longer interleave mid-reset. detect's publish path never takes the new mutex, so the unlock-before-walk latency improvement is preserved.

**Review by thiagoftsm — APPROVED** [review]
PR is working as expected, LGTM!
